File: sdk/python/seocheon/infrastructure/signing/vault.py
"""VaultSigningService: signs transactions via an external vault server (production)."""

from __future__ import annotations

import aiohttp


class VaultSigningService:
    """Signs transactions via an external vault server."""

    def __init__(self, endpoint: str, key_name: str) -> None:
        if not endpoint or not key_name:
            raise ValueError("vault endpoint and key name are required")
        self._endpoint = endpoint.rstrip("/")
        self._key_name = key_name
        self._address: str = ""
        self._pub_key: bytes = b""

    async def initialize(self) -> None:
        """Connect to the vault server to fetch address and public key."""
        timeout = aiohttp.ClientTimeout(total=10)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            # Fetch address
            addr_url = f"{self._endpoint}/v1/keys/{self._key_name}/address"
            async with session.get(addr_url) as resp:
                if resp.status != 200:
                    body = await resp.text()
                    raise RuntimeError(f"vault address request failed: {body}")
                data = await resp.json()
                self._address = data["address"]

            # Fetch public key
            pk_url = f"{self._endpoint}/v1/keys/{self._key_name}/pubkey"
            async with session.get(pk_url) as resp:
                if resp.status != 200:
                    body = await resp.text()
                    raise RuntimeError(f"vault pubkey request failed: {body}")
                data = await resp.json()
                self._pub_key = bytes.fromhex(data["pubkey"])

    def sign(self, tx_bytes: bytes) -> bytes:
        """Sign data via vault. Must call initialize() first in an async context.

        For synchronous use, call sign_sync. For async, call sign_async.
        """
        raise NotImplementedError("Use sign_async for vault signing")

    async def sign_async(self, tx_bytes: bytes) -> bytes:
        """Sign data via the vault server asynchronously."""
        timeout = aiohttp.ClientTimeout(total=10)
        url = f"{self._endpoint}/v1/keys/{self._key_name}/sign"
        payload = {"data": tx_bytes.hex()}

        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.post(url, json=payload) as resp:
                if resp.status != 200:
                    body = await resp.text()
                    raise RuntimeError(f"vault sign failed with status {resp.status}: {body}")
                data = await resp.json()
                return bytes.fromhex(data["signature"])

    def get_address(self) -> str:
        return self._address

    def get_pub_key(self) -> bytes:
        return self._pub_key

File: sdk/go/internal/signing/vault.go
package signing

import (
	"bytes"
	"context"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// vaultSignRequest is the JSON body for POST /v1/keys/{name}/sign.
type vaultSignRequest struct {
	Data string `json:"data"`
}

// vaultSignResponse is the JSON response from POST /v1/keys/{name}/sign.
type vaultSignResponse struct {
	Signature string `json:"signature"`
}

// vaultAddressResponse is the JSON response from GET /v1/keys/{name}/address.
type vaultAddressResponse struct {
	Address string `json:"address"`
}

// vaultPubKeyResponse is the JSON response from GET /v1/keys/{name}/pubkey.
type vaultPubKeyResponse struct {
	PubKey string `json:"pubkey"`
}

// initVaultService initializes the VaultService by fetching address and pubkey
// from the vault server.
func initVaultService(endpoint, keyName string) (*VaultService, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	client := &http.Client{Timeout: 10 * time.Second}
	baseURL := strings.TrimRight(endpoint, "/")

	// Fetch address
	addrURL := fmt.Sprintf("%s/v1/keys/%s/address", baseURL, keyName)
	addrResp, err := doVaultGet(ctx, client, addrURL)
	if err != nil {
		return nil, fmt.Errorf("fetching vault address: %w", err)
	}

	var addrResult vaultAddressResponse
	if err := json.Unmarshal(addrResp, &addrResult); err != nil {
		return nil, fmt.Errorf("parsing vault address response: %w", err)
	}

	// Fetch public key
	pubKeyURL := fmt.Sprintf("%s/v1/keys/%s/pubkey", baseURL, keyName)
	pubKeyResp, err := doVaultGet(ctx, client, pubKeyURL)
	if err != nil {
		return nil, fmt.Errorf("fetching vault pubkey: %w", err)
	}

	var pubKeyResult vaultPubKeyResponse
	if err := json.Unmarshal(pubKeyResp, &pubKeyResult); err != nil {
		return nil, fmt.Errorf("parsing vault pubkey response: %w", err)
	}

	pubKeyBytes, err := hex.DecodeString(pubKeyResult.PubKey)
	if err != nil {
		return nil, fmt.Errorf("decoding vault pubkey hex: %w", err)
	}

	return &VaultService{
		endpoint: baseURL,
		keyName:  keyName,
		address:  addrResult.Address,
		pubKey:   pubKeyBytes,
		client:   client,
	}, nil
}

// vaultSign sends a sign request to the vault server.
func vaultSign(client *http.Client, endpoint, keyName string, data []byte) ([]byte, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	url := fmt.Sprintf("%s/v1/keys/%s/sign", endpoint, keyName)

	reqBody := vaultSignRequest{
		Data: hex.EncodeToString(data),
	}
	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("marshaling sign request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("creating sign request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("sending sign request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("reading sign response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("vault sign failed with status %d: %s", resp.StatusCode, string(respBody))
	}

	var signResult vaultSignResponse
	if err := json.Unmarshal(respBody, &signResult); err != nil {
		return nil, fmt.Errorf("parsing sign response: %w", err)
	}

	sigBytes, err := hex.DecodeString(signResult.Signature)
	if err != nil {
		return nil, fmt.Errorf("decoding signature hex: %w", err)
	}

	return sigBytes, nil
}

// doVaultGet performs a GET request to the vault server.
func doVaultGet(ctx context.Context, client *http.Client, url string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("vault request failed with status %d: %s", resp.StatusCode, string(body))
	}

	return body, nil
}
